### PR TITLE
Bug fixes; reconnection

### PR DESCRIPTION
**Pull request by flcl42**
-  Improve yamux
-  Add reconnection to pubsub
-  Improve channels

### CODE REVIEW

File: src/libp2p/Directory.Packages.props [hunk 1]
```diff
@@ -5,7 +5,7 @@
   <ItemGroup>
     <PackageVersion Include="BenchmarkDotNet" Version="0.13.11" />
     <PackageVersion Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.13.11" />
-    <PackageVersion Include="BouncyCastle.Cryptography" Version="2.2.1" />
+    <PackageVersion Include="BouncyCastle.Cryptography" Version="2.3.1" />
```
Comment (rubo) [inline src/libp2p/Directory.Packages.props:8]: Why not the latest version?

File: src/libp2p/Libp2p.Core.TestsBase/TestChannel.cs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 // SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
 // SPDX-License-Identifier: MIT
 
+using Newtonsoft.Json.Linq;
```
Comment (rubo) [inline src/libp2p/Libp2p.Core.TestsBase/TestChannel.cs:4]: Do we really need this?

File: src/libp2p/Directory.Packages.props (hunk 1 above)
Comment (flcl42) [reply src/libp2p/Directory.Packages.props:8]: Just to make it the same as for other dependencies for now. We can upgrade it separately with packaging, wdyt?

File: src/libp2p/Libp2p.Core.TestsBase/TestChannel.cs (hunk 2 above)
Comment (flcl42) [reply src/libp2p/Libp2p.Core.TestsBase/TestChannel.cs]: Nope, removed with some other unused usings

**Review by rubo — APPROVED** [review]

**Pull request merged**